libs/ts-rest/nest/src/lib/ts-rest-nest-handler.ts

```diff
@@ -0,0 +1,374 @@
+import { Reflector } from '@nestjs/core';
+import { Observable, map } from 'rxjs';
+import type { Response, Request } from 'express-serve-static-core';
+import {
+  All,
+  SetMetadata,
+  UseInterceptors,
+  applyDecorators,
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  BadRequestException,
+  NotFoundException,
+  Get,
+  Post,
+  Put,
+  Patch,
+  Delete,
+  InternalServerErrorException,
+  HttpException,
+} from '@nestjs/common';
+import {
+  AppRouter,
+  isAppRoute,
+  AppRoute,
+  checkZodSchema,
+  parseJsonQueryObject,
+  ServerInferResponses,
+  isAppRouteOtherResponse,
+} from '@ts-rest/core';
+import {
+  JsonQuerySymbol,
+  TsRestAppRouteMetadataKey,
+  TsRestAppRouterMetadataKey,
+  ValidateResponsesSymbol,
+} from './constants';
+import { TsRestRequestShape } from './ts-rest-request.decorator';
+import { z } from 'zod';
+import { TsRestOptions } from './ts-rest.decorator';
+import { JsonQuery } from './json-query.decorator';
+
+export class RequestValidationError extends BadRequestException {
+  constructor(
+    public pathParams: z.ZodError | null,
+    public headers: z.ZodError | null,
+    public query: z.ZodError | null,
+    public body: z.ZodError | null
+  ) {
+    super({
+      paramsResult: pathParams,
+      headersResult: headers,
+      queryResult: query,
+      bodyResult: body,
+    });
+  }
+}
+
+export class ResponseValidationError extends InternalServerErrorException {
+  constructor(appRoute: AppRoute, error: z.ZodError) {
```

Comment:
This could help if someone wants to custom handle these using an exception filter

```suggestion
  constructor(public appRoute: AppRoute, public error: z.ZodError) {
```